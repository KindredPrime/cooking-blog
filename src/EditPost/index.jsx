import { Component } from 'react';
import PropTypes from 'prop-types';
import APIError from '../APIError/index';

class EditPost extends Component {
  state = {
    content: this.props.content,
    isSubmitting: false,
    error: null
  };

  handleCancel(e) {
    e.preventDefault();

    this.props.handleCancel();
  }

  handleSubmit(e) {
    e.preventDefault();

    const { content } = this.state;

    this.setState({
      isSubmitting: true
    });

    this.props.handleSubmit(content)
      .then(() => this.setState({
        isSubmitting: false,
        error: null
      }))
      .catch((error) => this.setState({
        isSubmitting: false,
        error
      }));
  }

  render() {
    const { content, isSubmitting, error } = this.state;

    return (
      <div className="EditPost">
        <form onSubmit={(e) => this.handleSubmit(e)}>
          <textarea
            defaultValue={content}
            onChange={(e) => this.setState({
              content: e.target.value
            })}
          ></textarea>

          <div className="EditPost__buttons">
            <button
              type="submit"
              disabled={isSubmitting}
            >
              Submit
            </button>
            <button
              type="submit"
              onClick={(e) => this.handleCancel(e)}
              disabled={isSubmitting}
            >
              Cancel
            </button>
          </div>
        </form>

        {isSubmitting && <p>Editing post...</p>}

        {error && <APIError message={error.message} />}
      </div>
    );
  }
}

EditPost.propTypes = {
  content: PropTypes.string.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  handleCancel: PropTypes.func.isRequired
};

export default EditPost;