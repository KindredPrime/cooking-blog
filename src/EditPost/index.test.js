import ReactDOM from 'react-dom';
import { render, screen, waitFor } from '@testing-library/react';
import UserEvent from '@testing-library/user-event';
import EditPost from './index';

describe('EditPost Component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <EditPost content="" handleSubmit={() => {}} handleCancel={() => {}} />,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders the UI as expected', () => {
    render(
      <EditPost
        content="Content of the post to be edited"
        handleSubmit={() => {}}
        handleCancel={() => {}}
      />);

    expect(document.body).toMatchSnapshot();
  });

  it('renders an error as expected', async () => {
    const error = 'Test error';
    render(
      <EditPost
        content="Content of the post to be edited"
        handleSubmit={() => new Promise((resolve, reject) => {
          throw new Error(error);
        })}
        handleCancel={() => {}}
      />
    );

    UserEvent.click(screen.getByText('Submit'));

    await waitFor(() => expect(screen.getByText(error)).toBeInTheDocument());
    expect(document.body).toMatchSnapshot();
  });
});