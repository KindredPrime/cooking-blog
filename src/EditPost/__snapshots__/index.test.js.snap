// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditPost Component renders an error as expected 1`] = `
<body>
  <div>
    <div
      class="EditPost"
    >
      <form>
        <textarea>
          Content of the post to be edited
        </textarea>
        <div
          class="EditPost__buttons"
        >
          <button
            type="submit"
          >
            Submit
          </button>
          <button
            type="submit"
          >
            Cancel
          </button>
        </div>
      </form>
      <p
        class="APIError"
      >
        Test error
      </p>
    </div>
  </div>
</body>
`;

exports[`EditPost Component renders the UI as expected 1`] = `
<body>
  <div>
    <div
      class="EditPost"
    >
      <form>
        <textarea>
          Content of the post to be edited
        </textarea>
        <div
          class="EditPost__buttons"
        >
          <button
            type="submit"
          >
            Submit
          </button>
          <button
            type="submit"
          >
            Cancel
          </button>
        </div>
      </form>
    </div>
  </div>
</body>
`;
